source/gateway.go

```diff
@@ -327,78 +408,35 @@ func (c *gatewayRouteResolver) resolve(rt gatewayRoute) (map[string]endpoint.Tar
 				continue
 			}
 			// Confirm that the Listener and Route ports match, if specified.
-			// EXPERIMENTAL: https://gateway-api.sigs.k8s.io/geps/gep-957/
 			if ref.Port != nil && *ref.Port != lis.Port {
 				continue
 			}
 			// Confirm that the Listener allows the Route (based on namespace and kind).
 			if !c.routeIsAllowed(gw.gateway, lis, rt) {
 				continue
 			}
-			// Find all overlapping hostnames between the Route and Listener.
-			// For {TCP,UDP}Routes, all annotation-generated hostnames should match since the Listener doesn't specify a hostname.
-			// For {HTTP,TLS}Routes, hostnames (including any annotation-generated) will be required to match any Listeners specified hostname.
-			gwHost := ""
-			if lis.Hostname != nil {
-				gwHost = string(*lis.Hostname)
-			}
-			for _, rtHost := range rtHosts {
-				if gwHost == "" && rtHost == "" {
-					// For {HTTP,TLS}Routes, this means the Route and the Listener both allow _any_ hostnames.
-					// For {TCP,UDP}Routes, this should always happen since neither specifies hostnames.
-					continue
-				}
-				host, ok := gwMatchingHost(gwHost, rtHost)
-				if !ok {
-					continue
-				}
-				override := getTargetsFromTargetAnnotation(gw.gateway.Annotations)
-				hostTargets[host] = append(hostTargets[host], override...)
-				if len(override) == 0 {
-					for _, addr := range gw.gateway.Status.Addresses {
-						hostTargets[host] = append(hostTargets[host], addr.Value)
+			// Confirm that the Listener and Route hostnames overlap, if specified.
+			overlap := true
+			if len(rt.Hostnames()) != 0 {
+				gwHost := getVal(lis.Hostname, "")
+				for _, rtHost := range rt.Hostnames() {
+					_, overlap = gwMatchingHost(string(gwHost), string(rtHost))
+					if overlap {
+						break
 					}
 				}
-				match = true
 			}
+			if !overlap {
+				continue
+			}
+			gwListeners[gw.gateway] = append(gwListeners[gw.gateway], lis)
+			match = true
 		}
 		if !match {
-			log.Debugf("Gateway %s/%s section %q does not match %s %s/%s hostnames %q", namespace, ref.Name, section, c.src.rtKind, meta.Namespace, meta.Name, rtHosts)
+			log.Debugf("Gateway %s/%s section %q does not match %s %s/%s hostnames %q", namespace, ref.Name, section, c.src.rtKind, meta.Namespace, meta.Name, rt.Hostnames())
```

Comment:
```suggestion
			log.Warnf("Gateway %s/%s section %q does not match %s %s/%s hostnames %q", namespace, ref.Name, section, c.src.rtKind, meta.Namespace, meta.Name, rt.Hostnames())
```
Shouldn't this be a warning ?

Reply:
Quite possibly; this comment was merely relocated from the [pre-existing sources](https://github.com/kubernetes-sigs/external-dns/blob/a61780280de4186fb5edb6fbbf7a36816c6910d3/source/gateway.go#L366)…

I'm not familiar with the desired log verbosity and was attempting to minimise _unnecessary_ changes in both behaviour and implementation.

I'll try to deduce what, if anything, that log is attempting to convey with a view to either _upgrading_ it or eliminating it.

Reply:
I see no compelling reason to change this logging…

It's arguably _debug_ level algorithm diagnostics, similar to several other `log.Debugf(…)` function calls throughout the pre-existing implementation. The algorithm in question is broadly a combination of the logic outlined beneath the [GatewaySpec.listeners](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.GatewaySpec) and [HTTPRouteSpec.hostnames](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRouteSpec) specifications. This particular log line appears to pertain to a scenario whereby a …Route object has been [accepted](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.GatewayClassConditionType) by a Gateway controller implementation (i.e. there is an affirmative entry in the [RouteStatus.parents](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.RouteStatus) list) but that _this_ implementation of the algorithm logic does not consider there to be a _match_ between the parent Gateway's listeners and the …Route in question.

One would perhaps not expect such a scenario to occur. If such a scenario does occur then it _might_ be a consequence of an issue with the Gateway controller or an issue with _this_ implementation.

I'm not particularly familiar with the log level semantics employed by external-dns, or Kubernetes as a whole, but I tend to employ:
* Error - something that is _never_ expected to happen and is generally catastrophic
* Warn - something that is expected to happen rarely and is potentially problematic
* Debug - something that is potentially of interest to someone investigating in detail

A case _could_ be made for upgrading most of the existing _debug_ level logging to _warn_ but, given the prior art, a log level of _debug_ seems reasonable to me in this instance.

Reply:
> I'm not particularly familiar with the log level semantics employed by external-dns, or Kubernetes as a whole, but I tend to employ:
>     * Error - something that is _never_ expected to happen and is generally catastrophic
>     * Warn - something that is expected to happen rarely and is potentially problematic
>     * Debug - something that is potentially of interest to someone investigating in detail

100% aligned with that.

> A case could be made for upgrading most of the existing debug level logging to warn but, given the prior art, a log level of debug seems reasonable to me in this instance.

Works for me, even if we should not restrain ourselves from the current prior art.
@ivankatliarchuk wdyt ?

Reply:
Debug is fine. I usually find myselft silencing WARN level as could not action on them